feat: Add GenWealth Demo App.

paulramsey opened the pull request:
# Description

Contributing the GenWealth Demo App, originally developed here: https://github.com/paulramsey/genwealth

Thank you for opening a Pull Request!
Before submitting your PR, there are a few things you can do to make sure it goes smoothly:

- [X] Follow the [`CONTRIBUTING` Guide](https://github.com/GoogleCloudPlatform/generative-ai/blob/main/CONTRIBUTING.md).
- [X] You are listed as the author in your notebook or README file.
  - [X] Your account is listed in [`CODEOWNERS`](https://github.com/GoogleCloudPlatform/generative-ai/blob/main/.github/CODEOWNERS) for the file(s).
- [X] Make your Pull Request title in the <https://www.conventionalcommits.org/> specification.
- [X] Ensure the tests and linter pass (Run `nox -s format` from the repository root to format).
- [X] Appropriate docs were updated (if necessary)

holtskinner commented on language/sample-apps/text-genwealth/images/genwealth-database-schema.png:
I would recommend moving all of these images and sample PDF files into a Public Google Cloud Storage bucket. We have one for this repo at https://console.cloud.google.com/storage/browser/github-repo
`gs://github-repo`

I can upload these files if you'd like to add them

paulramsey replied:
Moved these as requested.

holtskinner commented on language/sample-apps/text-genwealth/ui/.vscode/tasks.json:
Should these .vscode files be committed, or added to a .gitignore?

holtskinner commented on language/sample-apps/text-genwealth/ui/.editorconfig:
Should this be committed?

holtskinner commented on language/sample-apps/text-genwealth/.gitignore:
Can you add these to the main .gitignore for the repo?

holtskinner reviewed (changes requested):
Thanks for adding this! Please resolve the lint errors, and I made a few other comments

holtskinner commented:
@paulramsey Since this uses Gemini, I'd recommend moving this sample app into this directory: https://github.com/GoogleCloudPlatform/generative-ai/tree/main/gemini/sample-apps

paulramsey commented:
@holtskinner - I moved the GenWealth sample app into the gemini/ folder as requested. 

I also cleaned up the lion's share of linting errors, but there are 3 that I'm still having trouble resolving:
1. This is valid SCSS, and I'm not sure how to resolve the "unknown word" error:
        2024-04-03 19:15:37 [ERROR]   Found errors in [stylelint] linter!
        2024-04-03 19:15:37 [ERROR]   Error code: 2. Command output:
        ------
        /github/workspace/gemini/sample-apps/genwealth/ui/src/styles.scss: you should use the "customSyntax" option when linting something other than CSS
        gemini/sample-apps/genwealth/ui/src/styles.scss
         10:4  ✖  Unknown word  CssSyntaxError
        1 problem (1 error, 0 warnings)
        ------
2. My code actually does expect a return value from this function, so I'm not sure how to proceed with this one:
        2024-04-03 19:17:39 [ERROR]   Found errors in [mypy] linter!
        2024-04-03 19:17:39 [ERROR]   Error code: 1. Command output:
        ------
        gemini/sample-apps/genwealth/function-scripts/process-pdf/main.py:123: error: No return value expected  [return-value]
        Found 1 error in 1 file (checked 1 source file)
        ------
        
3. The sqlfluff linter is complaining about long lines, whitespace, and the inability for it to scan valid SQL code: 
        2024-04-03 19:17:41 [ERROR]   Found errors in [sqlfluff] linter!
        2024-04-03 19:17:41 [ERROR]   Error code: 1. Command output:
        ------
        == [/github/workspace/gemini/sample-apps/genwealth/database-files/genwealth-demo_llm.sql] FAIL
        L:   2 | P:   1 | LT05 | Line is too long (84 > [80](https://github.com/GoogleCloudPlatform/generative-ai/actions/runs/8544264671/job/23409897371?pr=480#step:4:81)). [layout.long_lines]
        L:   8 | P:   2 | LT01 | Unnecessary trailing whitespace. [layout.spacing]
        L:  10 | P:   2 | LT01 | Unnecessary trailing whitespace. [layout.spacing]
        L:  16 | P:   1 | LT05 | Line is too long (84 > 80). [layout.long_lines]
        L:  26 | P:   1 |  PRS | Line 26, Position 1: Found unparsable section: 'CREATE
                               | FUNCTION llm(

holtskinner commented:
I tried adding some custom configs for sqlfluff and stylelint to resolve the lint errors.

holtskinner commented on gemini/sample-apps/genwealth/function-scripts/write-metadata/main.py:
```diff
@@ -0,0 +1,63 @@
+    target_file_name = "{}.jsonl".format(ticker)
+
+    # Build metadata jsonl
+    metadata = (
```
Can this be done in a dictionary, then converted to jsonl? This is difficult to read.

paulramsey replied:
Will revisit this.

paulramsey replied:
Moving to backlog.

holtskinner commented on gemini/sample-apps/genwealth/function-scripts/write-metadata/main.py:
```diff
@@ -0,0 +1,63 @@
+    metadata_bucket = "{}-docs-metadata".format(project_id)
+    uid = str(uuid.uuid4())
+    ticker = Path(name).stem
+    target_file_name = "{}.jsonl".format(ticker)
```
Prefer f strings over `.format()`

paulramsey replied:
Fixing.

holtskinner commented on gemini/sample-apps/genwealth/function-scripts/analyze-prospectus/main.py:
```diff
@@ -0,0 +1,151 @@
+
+    # Prep model and template
+    model = VertexAI(
+        model_name="text-bison@002", max_output_tokens=1024, temperature=0.0
```
Could this use Gemini instead of text-bison?

paulramsey replied:
Yes, updated.

holtskinner commented on gemini/sample-apps/genwealth/function-scripts/analyze-prospectus/main.py:
```diff
@@ -0,0 +1,151 @@
+    )
+    template = """
+<MISSION>
+ You are an experiences financial analyst. Your task is to create a detailed
```
experienced

paulramsey replied:
Fixed.

holtskinner commented on gemini/sample-apps/genwealth/function-scripts/analyze-prospectus/main.py:
```diff
@@ -0,0 +1,151 @@
+        overview = "None"
+
+        for i in range(len(result)):
+            # print(result[i])
```
Remove commented out code

paulramsey replied:
Fixed.

holtskinner commented on gemini/sample-apps/genwealth/function-scripts/process-pdf/main.py:
```diff
@@ -0,0 +1,328 @@
+    project_id = os.environ["PROJECT_ID"]
+
+    # Doc AI Vars
+    source_file = "gs://{}/{}".format(bucket, name)
```
Same as above, use f strings instead of `.format()`

paulramsey replied:
Fixed.

holtskinner commented on gemini/sample-apps/genwealth/function-scripts/process-pdf/main.py:
```diff
@@ -0,0 +1,328 @@
+    processor_id = os.environ["PROCESSOR_ID"]  # Create processor before running sample
+
+    # AlloyDB Vars
+    cluster = "alloydb-cluster"
```
Declare these closer to the AlloyDB creation

paulramsey replied:
Fixed.

holtskinner reviewed (commented):
Added additional comments - Also what do you think about moving the images/pdfs to Cloud Storage instead of having them in this repo? The folder is already large as is.

paulramsey commented:
@holtskinner - Thanks for the feedback and the help with lint errors. 

Regarding the images and files, we can move those to a bucket. I'll just need to update lots of references. The images are used through the README's, and the sample files are automatically uploaded to the $PROJECT_ID-docs bucket as part of the deployment scripts. If you can upload them and provide the paths, I can work up updating the README ref's and the deployment code.

paulramsey commented on gemini/sample-apps/genwealth/function-scripts/process-pdf/main.py:
```diff
@@ -0,0 +1,328 @@
+            overwrite_existing=True,
+        )
+
+    # TO DO: Create vector index
```
Yes, will update.

paulramsey commented on gemini/sample-apps/genwealth/function-scripts/process-pdf/main.py:
```diff
@@ -0,0 +1,328 @@
+    # Send message to pubsub topic to kick off next step
+    ticker = Path(source_file).stem
+    publisher = pubsub_v1.PublisherClient()
+    topic_name = "projects/{}/topics/{}-doc-ready".format(project_id, project_id)
```
This is the way the the topic name is defined in the docs:
https://cloud.google.com/python/docs/reference/pubsub/latest

I'll update to use f strings to make it cleaner.

holtskinner commented:
Also, don't worry about the SQL/SCSS lint. I'm not sure how to configure it to override the default settings.

paulramsey commented on gemini/sample-apps/genwealth/function-scripts/process-pdf/main.py:
```diff
@@ -0,0 +1,328 @@
+
+
+# Source: https://cloud.google.com/document-ai/docs/samples/documentai-batch-process-document#documentai_batch_process_document-python
+def batch_process_documents(
```
I'll add this to the backlog to investigate.

holtskinner commented on gemini/sample-apps/genwealth/deployment/deploy-search.sh:
```diff
@@ -40,6 +40,9 @@ DATA_STORE_ID=$(curl -X GET \
+gsutil -m cp -r \
+  "gs://github-repo/generative-ai/sample-apps/genwealth/sample-prospectus" \
+  .
 gsutil cp sample-prospectus/*.pdf gs://"${DOCS_BUCKET}"
```
You don't need to download these files locally, they can be transferred directly between buckets.

```
gsutil -m cp gs://github-repo/generative-ai/sample-apps/genwealth/sample-prospectus/*.pdf gs://"${DOCS_BUCKET}"
```

paulramsey replied:
Good call - fixed this.

paulramsey commented:
@holtskinner - Thanks for uploading the files. I updated everything to point there and removed the images and pdfs from the project. I also addressed most of your comments (and added a couple to the backlog). Lint checks are passing except for the two we've talked about. Please have another look and let me know if there are any other priority items for me to update before merging. Thanks!

holtskinner commented:
Looks good! Thanks for the refactoring/getting this ready.
